File: scripts/bed_to_coverage.R
library(GenomicRanges)
library(data.table)

## chip = import()
f = snakemake@input[[1]]

print("Reading data table")
if (Sys.info()["sysname"] == "Darwin"){
  cmd = paste0("gzcat ", f, " | cut -f 1-3,6")
} else {
  cmd = paste0("zcat ", f, " | cut -f 1-3,6")
}

print(cmd)
df = fread(cmd, header=FALSE, col.names=c("Chromosome", "Start", "End", "Strand"), stringsAsFactors=TRUE)

start.time <- Sys.time()
print("Starting to create GRanges")
gr = GRanges(seqnames = df$Chromosome, ranges = IRanges(start = df$Start, end = df$End), strand=df$Strand)

print("Starting to create Rles")

chip_list = coverage(gr)

end.time <- Sys.time()


time.taken <- end.time - start.time

time.taken <- as.numeric(time.taken, units="secs")

write(time.taken, snakemake@output[["timing"]])
capture.output(print(chip_list), file=snakemake@output[["preview"]])

File: scripts/bed_to_coverage.py
import pandas as pd



import ray
num_cores = int(snakemake.wildcards.num_cores)
if num_cores != 1:
    ray.init(num_cpus=num_cores)
else:
    ray.init(local_mode=True, num_cpus=1) # logging_level=logging.CRITICAL # local_mode=True


from pyranges import PyRanges
from pyrle import PyRles

from time import time
import datetime

chip = pd.read_table(snakemake.input[0], sep="\t", usecols=[0, 1, 2, 5],
                     header=None, names="Chromosome Start End Strand".split(),
                     dtype={"Chromosome": "category", "Strand": "category"})
start = time()
grles = PyRles(chip, stranded=True)
end = time()
total = end - start

total_dt = datetime.datetime.fromtimestamp(total)

minutes_seconds = total_dt.strftime('%-M.%-S.%f')

open(snakemake.output["timing"], "w+").write(minutes_seconds)

print(grles)
